### DISCUSSION TITLE
Validation based on previous attribute

### DISCUSSION THREAD

**Discussion by mfrata** [pydantic/pydantic#2351] [Question] [answered]
Hi. 👋🏾 

Is there a way to validate an attribute based on a previous one?

E.g:
```python
from pydantic import BaseModel, validator
from typing import Optional

class Customer(BaseModel):
    ...
    wantsJuice: bool
    typeOfJuice: Optional[str] = None
    @validator("typeOfJuice")
    def check_type_of_juice_is_defined_if_wants_juice(cls, val):
        # if wantsJuice is True, typeOfJuice must be defined
        # else, ignore value of typeOfJuice
```
I've tried to access the `cls` and check if `wantsJuice` was defined. But couldn't find its reference there.
Also, I've tried to pass both of them(`@validator("wantsJuice","typeOfJuice")`) as I saw [here](https://github.com/samuelcolvin/pydantic/blob/master/pydantic/class_validators.py#L50) but `check_type_of_juice_is_defined_if_wants_juice` is called twice for each one, instead of once for both of them.

Thanks for the awesome lib!! 🚀

**Comment by PrettyWood** [comment, marked as answer]
Hi @mfrata and welcome!

You can have a look at the [doc](https://pydantic-docs.helpmanual.io/usage/validators/), which contains an example with `values` that contain the values already checked. Furthermore you probably want to use `always=True` to check the default value (in your case `None`).
Putting everything together you get

```py
class Customer(BaseModel):
    wantsJuice: bool
    typeOfJuice: Optional[str] = None

    @validator("typeOfJuice", always=True)
    def check_type_of_juice_is_defined_if_wants_juice(cls, val, values):
        if values.get('wantsJuice') and val is None:
            raise ValueError('typeOfJuice can not be None')
        return val
```

Hope it helps 😃

**Comment by samuelcolvin** [comment]
First discussion 👌

**Comment by mfrata** [comment]
Hi @PrettyWood and @samuelcolvin. It worked like a charm! Thanks a lot. You guys rock!!
